common/numerated/src/iterators.rs

```diff
@@ -0,0 +1,330 @@
+        Interval::try_from(range).map(Into::into)
+    }
+}
+
+impl<T: Numerated> Iterator for IntervalIterator<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.map(|interval| {
+            let start = interval.start();
+            self.0 = interval.inc_start();
+            start
+        })
+    }
+}
+
+impl<T: Numerated + UpperBounded> IntervalIterator<T> {
+    /// Returns interval `start..start + len` if it's possible.
+    /// - if `len == None`, then it is supposed, that `len == T::Distance::max_value() + 1`.
+    /// - if `start + len - 1` is out of `T`, then returns [OutOfBoundsError].
+    /// - if `len` is zero, then returns empty interval.
+    pub fn new_with_len<S: Into<T::Bound>, L: Into<Option<T::Distance>>>(
+        start: S,
+        len: L,
+    ) -> Result<Self, OutOfBoundsError> {
+        match Interval::new_with_len(start, len) {
+            Ok(interval) => Ok(interval.into()),
+            Err(NewWithLenError::ZeroLen) => Ok(Self(None)),
+            Err(NewWithLenError::OutOfBounds(err)) => Err(err),
+        }
+    }
+}
+
+impl<T: Display> Display for IntervalIterator<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if let Some(interval) = &self.0 {
+            write!(f, "{}", interval)
+        } else {
+            write!(f, "∅")
+        }
+    }
+}
+
+/// Helper struct to iterate over intervals from `tree1`, which are not in `tree2`.
+///
+/// See also [`IntervalsTree::difference`].
+pub struct DifferenceIterator<T: Numerated, I: Iterator<Item = Interval<T>>> {
+    pub(crate) iter1: I,
+    pub(crate) iter2: I,
+    pub(crate) interval1: Option<Interval<T>>,
+    pub(crate) interval2: Option<Interval<T>>,
+}
+
+impl<T: Numerated, I: Iterator<Item = Interval<T>>> Iterator for DifferenceIterator<T, I> {
+    type Item = Interval<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let interval1 = if let Some(interval1) = self.interval1 {
+                interval1
+            } else {
+                let interval1 = self.iter1.next()?;
+                self.interval1 = Some(interval1);
+                interval1
+            };
+
+            let interval2 = if let Some(interval2) = self.interval2 {
+                interval2
+            } else if let Some(interval2) = self.iter2.next() {
+                interval2
+            } else {
+                self.interval1 = None;
+                return Some(interval1);
+            };
+
+            if interval2.end() < interval1.start() {
+                self.interval2 = None;
+                continue;
+            }
+
+            self.interval2 = Some(interval2);
+
+            if interval1.end() < interval2.start() {
+                self.interval1 = None;
+                return Some(interval1);
+            } else {
+                if let Some(new_start) = interval2.end().inc_if_lt(interval1.end()) {
```

Comment:
for these algos you can also introduce intervals api like: `interval1.difference(interval2)`;
consider `interval1.cross(interval2) -> bool` or maybe `.meet()`/`.touches()`;
otherwise you have A LOT of repeated code that's not readable, while task is pretty obvious.

Reply:
Consider `is_subset` like in other libraries

Reply:
I cannot see any more readable possible solutions. If you have some more specific ideas I would be glad to consider them